hpy/devel/src/runtime/ctx_type.c

```diff
@@ -1130,6 +1136,18 @@ ctx_Type_FromSpec(HPyContext *ctx, HPyType_Spec *hpyspec,
         return HPy_NULL;
     }
 
+#if PY_VERSION_HEX >= 0x030B0000
+    /* Since Python 3.11, the name will be copied into a fresh buffer such that
+       the type objects owns this buffer. Hence we need to patch the pointer
+       again. The nice thing is that Python will then free out HPyType_Extra
+       for us. The internal field '_ht_tpname' is only used for freeing. So, we
+       can set it to 'extra'. */
+    PyHeapTypeObject *ht = (PyHeapTypeObject *) result;
+    PyMem_Free(ht->_ht_tpname);
+    ht->_ht_tpname = (char *) extra;
+    ht->ht_type.tp_name = (const char *)extra->name;
+#endif
```

Comment:
wow, this is brilliant.
It's also a bit scary, because if someone by chance tries to access _ht_tpname you will get nonsense. But it seems that nobody is accessing it anyway, so I suppose it's fine :+1:

Reply:
I agree. I checked the CPython code and it's really just used for freeing.
As I mentioned, we could maybe use a metaclass for HPy types to store the extra info. I'll investigate.